### 1. Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks): zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110], ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]. Используя математические операции, посчитать коэффициенты линейной регрессии, приняв за X заработную плату (то есть, zp - признак), а за y - значения скорингового балла (то есть, ks - целевая переменная). Произвести расчет как с использованием intercept, так и без.

In [5]:
import numpy as np
X=np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
y=np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
b=(np.mean(X*y)-np.mean(X)*np.mean(y))/(np.mean(X**2)-np.mean(X)**2)
a=np.mean(y)-b*np.mean(X)
y_hat=a+b*X
y_hat

array([535.89621821, 562.10160703, 942.07974498, 968.2851338 ,
       548.99891262, 627.61507909, 585.68645697, 837.25818968,
       758.64202321, 732.43663439])

In [6]:
a

444.1773573243596

In [7]:
b

2.620538882402765

Ответ: a= 444.1773573243596, b=2.620538882402765

_______________________________________________________________________________________________________________________________

### 2. Посчитать коэффициент линейной регрессии при заработной плате (zp), используя градиентный спуск (без intercept).

In [8]:
def mse_(B1,y=y,X=X,n=10):
    return np.sum((B1*X-y)**2)/n

In [9]:
alpha = 1e-6

In [10]:
B1=0.5
n=10

In [11]:
for i in range(10):
    B1 -=alpha*(2/n)*np.sum((B1*X-y)*X)
    print('B1={}'.format(B1))

B1=0.6485068
B1=0.7929217624382401
B1=0.9333576305334268
B1=1.0699240410678132
B1=1.2027276099794635
B1=1.3318720155961774
B1=1.4574580795760528
B1=1.579583845617878
B1=1.6983446560027995
B1=1.8138332260270231


In [12]:
for i in range(3000):
    B1 -=alpha*(2/n)*np.sum((B1*X-y)*X)
    if i%500==0:
        print('Iteration: {i}, B1={B1}, mse={mse}'.format(i=i, B1=B1, mse=mse_(B1)))

Iteration: 0, B1=1.9261397163836553, mse=272957.7796748012
Iteration: 500, B1=5.8898170236888845, mse=56516.858415878334
Iteration: 1000, B1=5.889820420129778, mse=56516.8584157194
Iteration: 1500, B1=5.889820420132673, mse=56516.85841571943
Iteration: 2000, B1=5.889820420132673, mse=56516.85841571943
Iteration: 2500, B1=5.889820420132673, mse=56516.85841571943


Ответ: b = 5.8898204

_______________________________________________________________________________________________________________________________

### 3. В каких случаях для вычисления доверительных интервалов и проверки статистических гипотез используется таблица значений функции Лапласа, а в каких - таблица критических точек распределения Стьюдента? 

Ответ: Если известна сигма, то используется таблица значений функции Лапласа, если сигма неизвестна используем таблицу критических точек распределения Стьюдента